### PR TITLE
leaderelection: Keep SwitchMetric for compatibility

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind cleanup
/kind api-change

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Mark interface `leaderelection.SwitchMetric` deprecated in favor of `leaderelection.LeaderMetric`
```

### CODE REVIEW

**Review by howardjohn — APPROVED** [review]

**Comment by dims** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/122069#discussion_r1563097713

thanks @howardjohn

**Comment by alvaroaleman** [comment]
/hold
If we want to wait for more opinions

File: staging/src/k8s.io/client-go/tools/leaderelection/metrics.go [hunk 1]
```diff
@@ -29,10 +29,17 @@ type leaderMetricsAdapter interface {
...
 
-// LeaderMetric instruments metrics used in leader election.
-type LeaderMetric interface {
+// SwitchMetric represents a single numerical value that can arbitrarily go up
+// and down.
+// Deprecated: Use LeaderMetric instead.
+type SwitchMetric interface {
 	On(name string)
 	Off(name string)
+}
```
Comment (pohly) [inline staging/src/k8s.io/client-go/tools/leaderelection/metrics.go:38]: Isn't the goal to restore compatibility with controller-runtime v0.17.0 (latest release)? Then this is not enough.

Code in controller-runtime which uses the `SwitchMetric` type compiles again, but this code here still fails:

https://github.com/kubernetes-sigs/controller-runtime/blob/c25fe2f5ca0f9e10d571f85da61c3d210d6345e7/pkg/metrics/leaderelection.go#L21

`leaderelectionMetricsProvider{}` is from controller-runtime and doesn't match what client-go's `leaderelection.SetProvider` expects:

```
pkg/metrics/leaderelection.go:21:29: cannot use leaderelectionMetricsProvider{} (value of type leaderelectionMetricsProvider) as leaderelection.MetricsProvider value in argument to leaderelection.SetProvider: leaderelectionMetricsProvider does not implement leaderelection.MetricsProvider (wrong type for method NewLeaderMetric)
		have NewLeaderMetric() leaderelection.SwitchMetric
		want NewLeaderMetric() leaderelection.LeaderMetric
```

If you want to make this change backward-compatible, then all of the APIs in client-go must continue to use `SwitchMetric`. In that place where you need to call `SlowpathExercised`, you have to do a checked cast of the `SwitchMetric` instance into `LeaderMetric` and only if you have one, then call `SlowpathExercised`.

To test, check out controller-runtime v0.17.0. `go get k8s.io/client-go@v0.30.0 && go mod vendor`, then edit `vendor/k8s.io/client-go/tools/metrics`.

Comment (howardjohn) [reply staging/src/k8s.io/client-go/tools/leaderelection/metrics.go:38]: I think there is also a middle ground where the change breaks v0.17.0, but its possible to ship a controller-runtime that works with client-go 0.29 and 0.30 (which, I _think_ this PR provides).
With the current state, users will need to atomically upgrade to some (future) v0.18 c-r and client-go at the same time.

I don't have enough context/stake in this to weigh the options, just throwing it out there. This has been done in the past with some breakages.

Comment (pohly) [reply staging/src/k8s.io/client-go/tools/leaderelection/metrics.go:38]: > its possible to ship a controller-runtime that works with client-go 0.29 and 0.30 (which, I think this PR provides)

I don't that works with this PR. [leaderelectionMetricsProvider.NewLeaderMetric](https://github.com/kubernetes-sigs/controller-runtime/blob/c25fe2f5ca0f9e10d571f85da61c3d210d6345e7/pkg/metrics/leaderelection.go#L26-L28) and thus controller-runtime can be compatible either with < 1.30 (returning SwitchMetric as it does in 0.7.0) or with >= 1.30 (returning the new LeaderMetric), but not both.

Comment (pohly) [reply staging/src/k8s.io/client-go/tools/leaderelection/metrics.go:38]: It's too late now for this PR. https://github.com/kubernetes-sigs/controller-runtime/pull/2693 already switch to the API from Kubernetes >= 1.30.

@linxiulei: Do agree to close it?

Comment (linxiulei) [reply staging/src/k8s.io/client-go/tools/leaderelection/metrics.go:38]: SG, thanks!

**Comment by linxiulei** [comment]
> It's too late now for this PR. https://github.com/kubernetes-sigs/controller-runtime/pull/2693 already switch to the API from Kubernetes >= 1.30.

**Pull request closed without merging**